AirCasting/Utils/Microphone/Calibration/MicrophoneCalibration.swift

```diff
@@ -0,0 +1,84 @@
+    /// - Returns: `Bool` value indicating wheter the calibration can be performed at the moment.
+    var isAvailable: Bool { get }
+    /// Starts a microphone calibration session
+    /// - Parameter completion: closure called when the calibation finishes
+    func startCalibration(completion: @escaping (Result<MicrophoneCalibrationDescription, MicrophoneCalibrationError>) -> ())
+}
+
+class MicrophoneCalibrator: MicrophoneCalibration {
+    private let minimalMeasurementsCount: UInt
+    private let calibrationDuration: Double
+    private let timeBetweenMeasurements: Double
+    
+    private let microphone: Microphone
+    private let dateProvider: () -> Date
+    @Injected private var timerScheduler: TimerScheduler
+    @Injected private var durationDecider: CalibrationDurationDecider
+    
+    var isAvailable: Bool { microphone.state == .notRecording }
+    
+    init(microphone: Microphone,
+         dateProvider: @escaping () -> Date = DateBuilder.getRawDate,
+         minimalMeasurementsCount: UInt = 12,
+         calibrationDuration: Double = 5.0,
+         timeBetweenMeasurements: Double = 0.25) {
+        self.microphone = microphone
+        self.dateProvider = dateProvider
+        self.minimalMeasurementsCount = minimalMeasurementsCount
+        self.calibrationDuration = calibrationDuration
+        self.timeBetweenMeasurements = timeBetweenMeasurements
+    }
+    
+    func startCalibration(completion: @escaping (Result<MicrophoneCalibrationDescription, MicrophoneCalibrationError>) -> ()) {
+        do {
+            try microphone.startRecording()
+        } catch {
+            completion(.failure(.microphoneError(error)))
+            return
+        }
+        var allMeasurements: [Double?] = []
+        let measurementStartDate = dateProvider()
+        var token: AnyObject!
+        token = timerScheduler.schedule(every: timeBetweenMeasurements) { [weak self, calibrationDuration, minimalMeasurementsCount] in
+            guard let self = self else { return }
+            allMeasurements.append(self.microphone.getCurrentDecibelLevel())
+            guard self.durationDecider.shouldCalibrationFinish(calibrationStartDate: measurementStartDate,
+                                                               desiredDuration: calibrationDuration) else { return }
+            try? self.microphone.stopRecording()
+            self.timerScheduler.stop(token: token)
+            let values = allMeasurements.compactMap { $0 }
+            guard values.count > minimalMeasurementsCount else {
+                completion(.failure(.couldntGetEnoughMeasurements))
+                return
+            }
+            let sortedValues = values.sorted(by: <)
+            completion(.success(.init(lowestPower: sortedValues.first!,
+                                      highestPower: sortedValues.last!)))
```

Comment:
I know it probably won't happen, but just catching out that you used `force unwrap` here

Reply:
here it's one of those occasions where (in my opinion) force unwrapping is ok. Reasons:
1. `sortedValues` originate from `values` which we check against the `minimalMeasurementsCount`
2. So it cannot ever be empty - we have 100% certainty that at least one element will be present
3. Final reason without which I think force unwrapping cannot be done: all of that is happening locally within a few lines of code - it's immediately obvious at first glance.